File: data_structures/linked_lists/doubly_linked_list.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.previous = null;
    }
}


class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    // O(N)
    get(index) {
        if (index < 0 || index >= this.length) {
            return null;
        }
        let node = this._getNodeAtIndex(index);
        return node.value;
    }

    // O(N)
    insert(index, value) {
        if (index < 0) {return this};

        let newNode = new Node(value);
        if (this._isEmpty()) {
            this.head = this.tail = newNode;
        }
        else if (index === 0) {
            return this.prepend(value);
        }
        else if (index >= this.length) {
            return this.append(value);
        }
        else {
            let leftNode = this._getNodeAtIndex(index - 1);
            let rightNode = leftNode.next;
            leftNode.next = newNode;
            newNode.next = rightNode;
            newNode.previous = leftNode;
            rightNode.previous = newNode;
        }
        this.length++;
        return this;
    }

    // O(N)
    delete(index) {
        if (index < 0 || index >= this.length) {return this};
        
        if (this._isEmpty()) {
            return this;
        }
        else if (index === 0){
            this.head = this.head.next;
            this.head.previous = null;
        }
        else {
            let leftNode = this._getNodeAtIndex(index - 1);
            let nodeToBeDeleted = leftNode.next;
            let rightNode = nodeToBeDeleted.next;
            leftNode.next = rightNode;
            rightNode.previous = leftNode;
            if (nodeToBeDeleted === this.tail) {
                this.tail = leftNode;
            }
        }
        this.length--;
        return this;
    }

    // O(N)
    search(value) {
        let position = 0;
        let currentNode = this.head;
        while (currentNode !== null) {
            if (currentNode.value === value) {
                return position;
            }
            currentNode = currentNode.next;
            position++;
        }
        return null;
    }

    prepend(value) {
        let node = new Node(value);
        if (this._isEmpty()) {
            this.tail = node;
        }
        node.next = this.head;
        this.head = node;
        this.length++;
        return this;
    }

    append(value) {
        let node = new Node(value);
        if (this._isEmpty()) {
            this.head = node;
        }
        else {
            this.tail.next = node;
        }
        node.previous = this.tail;
        this.tail = node;
        this.length++;
        return this;
    }

    printList() {
        const array = [];
        let currentNode = this.head;
        while(currentNode !== null){
            array.push(currentNode.value)
            currentNode = currentNode.next
        }
        console.log(array);
    }

    _isEmpty() {
        return this.head === null && this.tail === null;
    }

    _getNodeAtIndex(index) {
        let distanceFromHead = index - 1;
        let distanceFromTail = this.length - index;
        if (distanceFromHead < distanceFromTail) {
          position = 0;
          currentNode = this.head;
          while (position < index) {
            currentNode = currentNode.next;
            position++;
          }
          return currentNode;
        }
        else {
          position = this.length - 1;
          currentNode = this.tail;
          while (position > index) {
            currentNode = currentNode.previous;
            position--;
          }
          return currentNode;
        }
      }
}

File: data_structures/linked_lists/doubly_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.previous = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    # O(N)
    def get(self, index):
        if index < 0 or index >= self.length:
            return None
        node = self._get_node_at_index(index)
        return node.value

    # O(N)
    def insert(self, index, value):
        if index < 0:
            return self
        elif index >= self.length:
            return self.append(value)
        elif index == 0:
            return self.prepend(value)
        new_node = Node(value)
        if self._is_empty():
            self.head = self.tail = new_node
        else:
            left_node = self._get_node_at_index(index - 1)
            right_node = left_node.next
            left_node.next = new_node
            new_node.next = right_node
            new_node.previous = left_node
            right_node.previous = new_node
        self.length += 1
        return self

    # O(N)
    def delete(self, index):
        if index < 0 or index >= self.length:
            return self
        if self._is_empty():
            return self
        if index == 0:
            self.head = self.head.next
            self.head.previous = None
            self.head.next.previous = self.head.next
        else:
            left_node = self._get_node_at_index(index - 1)
            node_to_delete = left_node.next
            right_node = node_to_delete.next
            left_node.next = right_node
            right_node.previous = left_node
            if node_to_delete == self.tail:
                self.tail = left_node
        self.length -= 1
        return self

    # O(N)
    def search(self, value):
        position = 0
        current_node = self.head
        while current_node is not None:
            if current_node.value == value:
                return position
            current_node = current_node.next
            position += 1
        return -1

    def prepend(self, value):
        node = Node(value)
        if self._is_empty():
            self.tail = node
        node.next = self.head
        self.head = node
        self.length += 1
        return self

    def append(self, value):
        node = Node(value)
        if self._is_empty():
            self.head = node
        node.previous = self.tail
        self.tail.next = node
        self.tail = node
        self.length += 1
        return self

    def print_as_list(self):
        array = []
        current_node = self.head
        while current_node is not None:
            array.append(current_node.value)
            current_node = current_node.next
        print(array)

    def _is_empty(self):
        return self.head is None and self.tail is None

    def _get_node_at_index(self, index):
        distance_from_head = index - 1
        distance_from_tail = self.length - index
        if distance_from_head < distance_from_tail:
            position = 0
            current_node = self.head
            while position < index:
                current_node = current_node.next
                position += 1
            return current_node
        else:
            position = self.length - 1
            current_node = self.tail
            while position > index:
                current_node = current_node.previous
                position -= 1
            return current_node